commanduino/commandmanager.py

```diff
@@ -65,48 +65,98 @@ def __init__(self, command_configs, devices_dict, init_timeout=DEFAULT_INIT_TIME
+                self.logger.warning("Additional error message: %s", e)
+
+        # Initialize device
+        if handler is not None:
             try:
                 elapsed = self.wait_device_for_init(handler)
-                self.logger.info('Found CommandManager at %s, init time was %s seconds', device_name,  round(elapsed, 3))
+                self.logger.info('Found Arduino CommandManager at %s, init time was %s seconds', device_name, round(elapsed, 3))
             except InitError:
-                self.logger.warning('CommandManager at %s has not initialized', device_name)
+                self.logger.warning('Arduino CommandManager at %s has not initialized', device_name)
+                return
+            # Update handlers list
             self.commandhandlers.append(handler)
 
-        self.register_all_devices(devices_dict)
-        self.set_devices_as_attributes()
-        self.initialised = True
+    def remove_command_handler(self, handler_to_remove):
+        """
+        Deletes the command handler object & removes a reference to it from
+        class dictionary. Then deletes the devices bound to the handler being deleted.
+
+        The method to detect which devices depend on the handler being deleted
+        may seem wanky because it is.
+        However, it is the only possible method as every device internal
+        CommanHandler doesn't hold a reference to an actual command handler.
+        The only thing having this reference is the device's write() function
+        which gets updated on create_and_setup_device()
+        """
+        if handler_to_remove not in self.commandhandlers:
+            self.logger.warning("Command handler %s not found!", handler_to_remove)
+            return
+        self.logger.info("Removing devices...")
+        # Find the dependent devices to remove
+        for name, dev in self.devices.copy().items():
+            if handler_to_remove is dev.write.__self__:
+                self.logger.info("Removing dependent device %s", name)
+                self.unregister_device(name)
+        # Remove handler
+        self.logger.info("Removing command handler...")
+        self.commandhandlers.remove(handler_to_remove)
 
     def set_devices_as_attributes(self):
         """
         Sets the list of devices as attributes.
         """
         for device_name, device in list(self.devices.items()):
             if hasattr(self, device_name):
-                self.logger.warning("Device named {device_name} is already a reserved attribute, please change name or do not use this pump in attribute mode, rather use pumps[{device_name}]".format(device_name=device_name))
+                self.logger.warning("Device named {device_name} is already a reserved attribute, please change name or do not use this device in attribute mode, rather use devices[{device_name}]".format(device_name=device_name))
```

Comment:
You forgot an f for the f-string here, or logging does some magic?

Reply:
It's not an f-string, it's an old plain `format()`-based formatting.
If you compare the change, it's an old John's code, I just changed 'pump' to 'device' - that was an obvious type.